### PR TITLE
Implementation of java.time.Year

### PR DESCRIPTION
**Pull request by alonsodomin**
Submitting for review an implementation of `java.time.Year`.

### CODE REVIEW

**Comment by sjrd** [comment]
Jenkins, test this please

File: javalib/src/main/scala/java/time/Year.scala [hunk 1]
```diff
@@ -0,0 +1,203 @@
...
+  // Not implemented
+  // def format(formatter: DateTimeFormatter): String
+
+}
+
+object Year {
+  import ChronoField._
+
+  val MIN_VALUE: Int = -999999999
+  val MAX_VALUE: Int = 999999999
```
Comment (nicolasstucki) [inline javalib/src/main/scala/java/time/Year.scala]: Make both `MIN_VALUE` and `MAX_VALUE` `final val`s.

Comment (nicolasstucki) [inline javalib/src/main/scala/java/time/Year.scala]: Now that we have these values you should rewrite the definition of `ChronoField.YEAR` to use them. 
```scala
  final val YEAR = new ChronoField("Year", 26,
      ValueRange.of(Year.MIN_VALUE, Year.MAX_VALUE), YEARS, FOREVER, isDateBasedFlag)
```

This is to ensure coherence with `YEAR.checkValidIntValue(year)` bellow (l.190).

File: javalib/src/main/scala/java/time/Year.scala [hunk 2]
```diff
@@ -0,0 +1,203 @@
...
+
+  def isLeap(year: Long): Boolean = iso.isLeapYear(year)
+
+  def of(year: Int): Year = {
+    YEAR.checkValidIntValue(year)
+    new Year(year)
+  }
+
+  def from(temporal: TemporalAccessor): Year = temporal match {
+    case t: Year => t
```
Comment (nicolasstucki) [inline javalib/src/main/scala/java/time/Year.scala]: Use `case temporal: Year => t`. In general when you just refine the type of a field just rebind it to the same name.

File: javalib/src/main/scala/java/time/Year.scala [hunk 3]
```diff
@@ -0,0 +1,203 @@
...
+
+  private val iso = IsoChronology.INSTANCE
+
+  def now(): Year = from(LocalDate.now())
+
+  // Not implemented
+  // def now(zone: ZoneId): Year
+  // def now(clock: Clock): Year
+
+  def isLeap(year: Long): Boolean = iso.isLeapYear(year)
```
Comment (nicolasstucki) [inline javalib/src/main/scala/java/time/Year.scala]: This definition should go after `def parse`.

File: javalib/src/main/scala/java/time/Year.scala [hunk 4]
```diff
@@ -0,0 +1,203 @@
...
+
+  def isSupported(field: TemporalField): Boolean = field match {
+    case _: ChronoField => field == YEAR_OF_ERA || field == YEAR || field == ERA
+    case null           => false
+    case _              => field.isSupportedBy(this)
+  }
+
+  def isSupported(unit: TemporalUnit): Boolean = unit match {
+    case _: ChronoUnit => unit == YEARS || unit == DECADES || unit == CENTURIES ||
+      unit == MILLENNIA || unit == ERAS
```
Comment (nicolasstucki) [inline javalib/src/main/scala/java/time/Year.scala]: Start `unit == YEARS || ...` on the next line and add a new line before the next `case`.

File: javalib/src/main/scala/java/time/Year.scala [hunk 5]
```diff
@@ -0,0 +1,203 @@
...
+    case null          => false
+    case _             => unit.isSupportedBy(this)
+  }
+
+  override def range(field: TemporalField): ValueRange = field match {
+    case YEAR_OF_ERA =>
+      if (year <= 0) ValueRange.of(1, MAX_VALUE + 1)
+      else ValueRange.of(1, MAX_VALUE)
+    case _           => super.range(field)
```
Comment (nicolasstucki) [inline javalib/src/main/scala/java/time/Year.scala]: Put a new line before this `case _` and don't align the `=>`.

File: javalib/src/main/scala/java/time/Year.scala [hunk 6]
```diff
@@ -0,0 +1,203 @@
...
+
+  def getLong(field: TemporalField): Long = field match {
+    case YEAR_OF_ERA => if (year < 1) 1 - year else year
+    case YEAR        => year
+    case ERA         => if (year < 1) 0 else 1
+
+    case _: ChronoField =>
+      throw new UnsupportedTemporalTypeException(s"Unsupported field: $field")
+
+    case _ => field.getFrom(this)
```
Comment (nicolasstucki) [inline javalib/src/main/scala/java/time/Year.scala]: As the year is an `Int` you should be able to just do:
```scala
def getLong(field: TemporalField): Long = get(field).toLong
```

File: javalib/src/main/scala/java/time/Year.scala [hunk 7]
```diff
@@ -0,0 +1,203 @@
...
+  }
+
+  def isLeap(): Boolean = Year.isLeap(year)
+
+  def isValidMonthDay(monthDay: MonthDay): Boolean =
+    monthDay.isValidYear(year)
+
+  def length(): Int = if (isLeap()) 366 else 365
```
Comment (nicolasstucki) [inline javalib/src/main/scala/java/time/Year.scala]: Use `if (isLeap) ...`

File: javalib/src/main/scala/java/time/Year.scala [hunk 8]
```diff
@@ -0,0 +1,203 @@
...
+    field match {
+      case YEAR_OF_ERA =>
+        val yearOfEra = YEAR_OF_ERA.checkValidIntValue(value)
+        val newYear = if (year < 1) 1 - yearOfEra else yearOfEra
+        withYear(newYear)
+      case YEAR =>
+        withYear(YEAR.checkValidIntValue(value))
+      case ERA =>
+        requireDateTime(value == 0 || value == 1,
+          s"Invalid value for field $field: $value")
```
Comment (nicolasstucki) [inline javalib/src/main/scala/java/time/Year.scala]: 4 spaces indentation.

File: javalib/src/main/scala/java/time/Year.scala [hunk 9]
```diff
@@ -0,0 +1,203 @@
...
+      if (year != newYear) Year.of(newYear)
+      else this
+    }
+
+    field match {
+      case YEAR_OF_ERA =>
+        val yearOfEra = YEAR_OF_ERA.checkValidIntValue(value)
+        val newYear = if (year < 1) 1 - yearOfEra else yearOfEra
+        withYear(newYear)
+      case YEAR =>
```
Comment (nicolasstucki) [inline javalib/src/main/scala/java/time/Year.scala]: New line before the `case`.

File: javalib/src/main/scala/java/time/Year.scala [hunk 10]
```diff
@@ -0,0 +1,203 @@
...
+        withYear(YEAR.checkValidIntValue(value))
+      case ERA =>
```
Comment (nicolasstucki) [inline javalib/src/main/scala/java/time/Year.scala]: Same

File: javalib/src/main/scala/java/time/Year.scala [hunk 11]
```diff
@@ -0,0 +1,203 @@
...
+        requireDateTime(value == 0 || value == 1,
+          s"Invalid value for field $field: $value")
+        val era = getLong(ERA)
```
Comment (nicolasstucki) [inline javalib/src/main/scala/java/time/Year.scala]: Use `get(ERA)` as the results can only be `0` or `1`.

Comment (nicolasstucki) [inline javalib/src/main/scala/java/time/Year.scala]: And use `if (era == != value.toInt)`

File: javalib/src/main/scala/java/time/Year.scala [hunk 12]
```diff
@@ -0,0 +1,203 @@
...
+        field.adjustInto(this, value)
+    }
+  }
+
+  def plus(amount: Long, unit: TemporalUnit): Year = unit match {
+    case YEARS     => plusYears(amount)
+    case DECADES   => plusYears(MathJDK8Bridge.multiplyExact(amount, 10))
+    case CENTURIES => plusYears(MathJDK8Bridge.multiplyExact(amount, 100))
+    case MILLENNIA => plusYears(MathJDK8Bridge.multiplyExact(amount, 1000))
+    case ERAS      =>
```
Comment (nicolasstucki) [inline javalib/src/main/scala/java/time/Year.scala]: Add a new line before `case ERAS`

File: javalib/src/main/scala/java/time/Year.scala [hunk 13]
```diff
@@ -0,0 +1,203 @@
...
+  def until(end: Temporal, unit: TemporalUnit): Long = {
+    val other = Year.from(end)
+    val yearsDiff: Long = other.getValue.toLong - year
+
+    unit match {
+      case YEARS     => yearsDiff
+      case DECADES   => yearsDiff / 10
+      case CENTURIES => yearsDiff / 100
+      case MILLENNIA => yearsDiff / 1000
+      case ERAS      => other.getLong(ERA) - getLong(ERA)
```
Comment (nicolasstucki) [inline javalib/src/main/scala/java/time/Year.scala]: Use `other.get(ERA) - get(ERA)`.

File: javalib/src/main/scala/java/time/Year.scala [hunk 14]
```diff
@@ -0,0 +1,203 @@
...
+
+  // Not implemented
+  // def query[R](query: TemporalQuery[R]): R
+
+  def adjustInto(temporal: Temporal): Temporal =
+    temporal.`with`(YEAR, year)
+
+  def until(end: Temporal, unit: TemporalUnit): Long = {
+    val other = Year.from(end)
+    val yearsDiff: Long = other.getValue.toLong - year
```
Comment (nicolasstucki) [inline javalib/src/main/scala/java/time/Year.scala]: Use `Int` for `yearsDiff`.

File: javalib/src/main/scala/java/time/Year.scala [hunk 15]
```diff
@@ -0,0 +1,203 @@
...
+
+  def compareTo(other: Year): Int = year - other.getValue
+
+  def isAfter(other: Year): Boolean = compareTo(other) > 0
+
+  def isBefore(other: Year): Boolean = compareTo(other) < 0
+
+  override def equals(other: Any): Boolean = other match {
+    case that: Year => year == that.getValue
+    case _ => false
```
Comment (nicolasstucki) [inline javalib/src/main/scala/java/time/Year.scala]: Align the `=>`

File: javalib/src/main/scala/java/time/Year.scala [hunk 16]
```diff
@@ -0,0 +1,203 @@
...
+  }
+
+  override def hashCode(): Int = year.hashCode()
+
+  override def toString: String = s"$year"
```
Comment (nicolasstucki) [inline javalib/src/main/scala/java/time/Year.scala]: Use `year.toString`

File: javalib/src/main/scala/java/time/Year.scala [hunk 17]
```diff
@@ -0,0 +1,203 @@
...
+
+  // Not implemented
+  // def format(formatter: DateTimeFormatter): String
```
Comment (nicolasstucki) [inline javalib/src/main/scala/java/time/Year.scala]: This method should be after `until` and before `atDay `.

File: test-suite/shared/src/test/require-jdk8/org/scalajs/testsuite/javalib/time/YearTest.scala [hunk 18]
```diff
@@ -0,0 +1,329 @@
...
+      field <- ChronoField.values()
+    } {
+      if (isSupported(field)) {
+        val expected = (field match {
+          case YEAR_OF_ERA => if (t.getValue < 1) 1 - t.getValue else t.getValue
+          case YEAR        => t.getValue
+          case ERA         => if (t.getValue < 1) 0 else 1
+          case _           =>
+            // Should not happen
+            throw new UnsupportedTemporalTypeException(s"Unsupported field: $field")
```
Comment (nicolasstucki) [inline test-suite/shared/src/test/require-jdk8/org/scalajs/testsuite/javalib/time/YearTest.scala]: You can remove the `case _ =>` and let it fail with a `MatchError`.

File: test-suite/shared/src/test/require-jdk8/org/scalajs/testsuite/javalib/time/YearTest.scala [hunk 19]
```diff
@@ -0,0 +1,329 @@
...
+    assertEquals(0L, lastBCYear.getLong(ERA))
+
+    assertEquals(101L, negativeYear.getLong(YEAR_OF_ERA))
+    assertEquals(-100L, negativeYear.getLong(YEAR))
+    assertEquals(0L, negativeYear.getLong(ERA))
+  }
+
+  @Test def length(): Unit = {
+    assertEquals(366, leapYear.length())
+    assertEquals(365, nonLeapYear.length())
```
Comment (nicolasstucki) [inline test-suite/shared/src/test/require-jdk8/org/scalajs/testsuite/javalib/time/YearTest.scala]: And all the cases in `samples` one to make sure that it does not fail on the corner cases.

File: test-suite/shared/src/test/require-jdk8/org/scalajs/testsuite/javalib/time/YearTest.scala [hunk 20]
```diff
@@ -0,0 +1,329 @@
...
+  }
+
+  @Test def `with`(): Unit = {
+    assertEquals(lastBCYear, lastBCYear.`with`(YEAR, 0))
+    assertEquals(Year.of(0), lastBCYear.`with`(YEAR_OF_ERA, 1))
```
Comment (nicolasstucki) [inline test-suite/shared/src/test/require-jdk8/org/scalajs/testsuite/javalib/time/YearTest.scala]: Add also the case where `with(YEAR_OF_ERA, 0)`. Do the same for the other years bellow.

File: test-suite/shared/src/test/require-jdk8/org/scalajs/testsuite/javalib/time/YearTest.scala [hunk 21]
```diff
@@ -0,0 +1,329 @@
...
+    assertEquals(Year.of(-50), lastBCYear.`with`(YEAR, -50))
+    assertEquals(Year.of(50), lastBCYear.`with`(YEAR, 50))
+    assertEquals(Year.of(1), lastBCYear.`with`(ERA, 1))
```
Comment (nicolasstucki) [inline test-suite/shared/src/test/require-jdk8/org/scalajs/testsuite/javalib/time/YearTest.scala]: Add also the case where with(ERA, 1). Do the same for the other years bellow.
 Add a line note

File: test-suite/shared/src/test/require-jdk8/org/scalajs/testsuite/javalib/time/YearTest.scala [hunk 22]
```diff
@@ -0,0 +1,329 @@
...
+    for (t <- samples) {
+      for (v <- Seq(Long.MinValue, -1000000000L, 1000000000L, Long.MaxValue)) {
+        expectThrows(classOf[DateTimeException], t.`with`(YEAR, v))
+      }
+      for (v <- Seq(Long.MinValue, -1000000001L, 1000000001L, Long.MaxValue)) {
+        expectThrows(classOf[DateTimeException], t.`with`(YEAR_OF_ERA, v))
+      }
+      for (v <- Seq(Long.MinValue, -1L, 2L, Long.MaxValue)) {
+        expectThrows(classOf[DateTimeException], t.`with`(ERA, v))
+      }
```
Comment (nicolasstucki) [inline test-suite/shared/src/test/require-jdk8/org/scalajs/testsuite/javalib/time/YearTest.scala]: Unroll the internal `for` loops.

File: test-suite/shared/src/test/require-jdk8/org/scalajs/testsuite/javalib/time/YearTest.scala [hunk 23]
```diff
@@ -0,0 +1,329 @@
...
+    expectThrows(classOf[DateTimeException], firstACYear.plus(1, ERAS))
+    expectThrows(classOf[DateTimeException], lastBCYear.plus(-1, ERAS))
+  }
+
+  @Test def plusYears(): Unit = {
+    for (t <- samples) {
+      assertSame(t, t.plusYears(0))
+    }
+
+    assertEquals(firstACYear, lastBCYear.plusYears(1))
```
Comment (nicolasstucki) [inline test-suite/shared/src/test/require-jdk8/org/scalajs/testsuite/javalib/time/YearTest.scala]: Add a case where the input is negative.

File: test-suite/shared/src/test/require-jdk8/org/scalajs/testsuite/javalib/time/YearTest.scala [hunk 24]
```diff
@@ -0,0 +1,329 @@
...
+    expectThrows(classOf[DateTimeException], min.plusYears(-1))
+    expectThrows(classOf[DateTimeException], max.plusYears(1))
```
Comment (nicolasstucki) [inline test-suite/shared/src/test/require-jdk8/org/scalajs/testsuite/javalib/time/YearTest.scala]: Add some cases where the input of `plusYears` is outside the allowed bounds.

File: test-suite/shared/src/test/require-jdk8/org/scalajs/testsuite/javalib/time/YearTest.scala [hunk 25]
```diff
@@ -0,0 +1,329 @@
...
+  }
+
+  @Test def minusYears(): Unit = {
+    for (t <- samples) {
+      assertSame(t, t.minusYears(0))
+    }
+
+    assertEquals(lastBCYear, firstACYear.minusYears(1))
+    expectThrows(classOf[DateTimeException], min.minusYears(Long.MinValue))
+    expectThrows(classOf[DateTimeException], max.minusYears(Long.MaxValue))
```
Comment (nicolasstucki) [inline test-suite/shared/src/test/require-jdk8/org/scalajs/testsuite/javalib/time/YearTest.scala]: Same as with `plusYears`

File: test-suite/shared/src/test/require-jdk8/org/scalajs/testsuite/javalib/time/YearTest.scala [hunk 26]
```diff
@@ -0,0 +1,329 @@
...
+  @Test def plusYears(): Unit = {
+    for (t <- samples) {
+      assertSame(t, t.plusYears(0))
+    }
+
+    assertEquals(firstACYear, lastBCYear.plusYears(1))
+    expectThrows(classOf[DateTimeException], min.plusYears(-1))
+    expectThrows(classOf[DateTimeException], max.plusYears(1))
+  }
+
```
Comment (nicolasstucki) [inline test-suite/shared/src/test/require-jdk8/org/scalajs/testsuite/javalib/time/YearTest.scala]: Missing tests for `minus`.

File: test-suite/shared/src/test/require-jdk8/org/scalajs/testsuite/javalib/time/YearTest.scala [hunk 27]
```diff
@@ -0,0 +1,329 @@
...
+    }
+
+    assertEquals(lastBCYear, firstACYear.minusYears(1))
+    expectThrows(classOf[DateTimeException], min.minusYears(Long.MinValue))
+    expectThrows(classOf[DateTimeException], max.minusYears(Long.MaxValue))
+  }
+
+  @Test def adjustInto(): Unit = {
+    val aDate = LocalDate.of(2015, 1, 1)
+    assertEquals(LocalDate.of(0, 1, 1), lastBCYear.adjustInto(aDate))
```
Comment (nicolasstucki) [inline test-suite/shared/src/test/require-jdk8/org/scalajs/testsuite/javalib/time/YearTest.scala]: Add a case with leap years on the 29 of February.

Comment (nicolasstucki) [inline test-suite/shared/src/test/require-jdk8/org/scalajs/testsuite/javalib/time/YearTest.scala]: Add some other years from your `samples`

File: test-suite/shared/src/test/require-jdk8/org/scalajs/testsuite/javalib/time/YearTest.scala [hunk 28]
```diff
@@ -0,0 +1,329 @@
...
+    assertEquals(1999999998L, min.until(max, YEARS))
+    assertEquals(-1999999998L, max.until(min, YEARS))
+    assertEquals(1L, lastBCYear.until(firstACYear, YEARS))
+
+    val otherLeapYear: Year = {
+      var y = leapYear.getValue + 1
+      while (!Year.isLeap(y)) y += 1
+      Year.of(y)
+    }
+    assertEquals(4L, leapYear.until(otherLeapYear, YEARS))
```
Comment (nicolasstucki) [inline test-suite/shared/src/test/require-jdk8/org/scalajs/testsuite/javalib/time/YearTest.scala]: I would suggest using a tailrec function rather than a while.
```scala 
@tailrec def otherLeapYear(y: Int): Year = {
  val nextYear = y + 1
  if (Year.isLeap(nextYear)) Year.of(nextYear)
  else nextLeapYear(nextYear)
}
assertEquals(4L, leapYear.until(otherLeapYear(leapYear), YEARS))
```

File: test-suite/shared/src/test/require-jdk8/org/scalajs/testsuite/javalib/time/YearTest.scala [hunk 29]
```diff
@@ -0,0 +1,329 @@
...
+  }
+
+  @Test def atDay(): Unit = {
+    for (t <- samples) {
+      assertEquals(LocalDate.of(t.getValue, 1, 1), t.atDay(1))
+      assertEquals(LocalDate.of(t.getValue, 12, 31), t.atDay(t.length()))
+    }
+
+    assertEquals(LocalDate.of(leapYear.getValue, 2, 29), leapYear.atDay(31 + 29))
+    assertEquals(LocalDate.of(leapYear.getValue, 3, 1), leapYear.atDay(31 + 30))
```
Comment (nicolasstucki) [inline test-suite/shared/src/test/require-jdk8/org/scalajs/testsuite/javalib/time/YearTest.scala]: Add tests for `leapYear.atDay(366)`, `leapYear.atDay(367)`, `nonLeapYear.atDay(365)`, `nonLeapYear.atDay(366)` and `.atDay(-1)`.

File: test-suite/shared/src/test/require-jdk8/org/scalajs/testsuite/javalib/time/YearTest.scala [hunk 30]
```diff
@@ -0,0 +1,329 @@
...
+  }
+
+  @Test def atMonthDay(): Unit = {
+    val leapMonthDay = MonthDay.of(2, 29)
+    val leapDate = LocalDate.of(leapYear.getValue(),
+        leapMonthDay.getMonthValue, leapMonthDay.getDayOfMonth)
+    assertEquals(leapDate, leapYear.atMonthDay(leapMonthDay))
+
+    val nonLeapDate = LocalDate.of(nonLeapYear.getValue(),
+      leapMonthDay.getMonthValue, 28)
```
Comment (nicolasstucki) [inline test-suite/shared/src/test/require-jdk8/org/scalajs/testsuite/javalib/time/YearTest.scala]: 4 spaces indentation.

File: test-suite/shared/src/test/require-jdk8/org/scalajs/testsuite/javalib/time/YearTest.scala [hunk 31]
```diff
@@ -0,0 +1,329 @@
...
+    assertEquals(nonLeapDate, nonLeapYear.atMonthDay(leapMonthDay))
+
+    val otherMonthDays = Month.values().map(m => MonthDay.of(m, m.minLength()))
```
Comment (nicolasstucki) [inline test-suite/shared/src/test/require-jdk8/org/scalajs/testsuite/javalib/time/YearTest.scala]: Consider factoring this out of `atMonthDay` and `isValidMonthDay`.

File: test-suite/shared/src/test/require-jdk8/org/scalajs/testsuite/javalib/time/YearTest.scala [hunk 32]
```diff
@@ -0,0 +1,329 @@
...
+      t <- samples
+      monthDay <- otherMonthDays
+    } {
+      val expected = LocalDate.of(t.getValue, monthDay.getMonthValue, monthDay.getDayOfMonth)
+      assertEquals(expected, t.atMonthDay(monthDay))
+    }
+  }
+
+  @Test def compareTo(): Unit = {
+    assertTrue(min.compareTo(min) == 0)
```
Comment (nicolasstucki) [inline test-suite/shared/src/test/require-jdk8/org/scalajs/testsuite/javalib/time/YearTest.scala]: Use `assertEquals`

File: test-suite/shared/src/test/require-jdk8/org/scalajs/testsuite/javalib/time/YearTest.scala [hunk 33]
```diff
@@ -0,0 +1,329 @@
...
+    assertTrue(min.compareTo(max) < 0)
+    assertTrue(max.compareTo(min) > 0)
+    assertTrue(max.compareTo(max) == 0)
```
Comment (nicolasstucki) [inline test-suite/shared/src/test/require-jdk8/org/scalajs/testsuite/javalib/time/YearTest.scala]: Same

File: test-suite/shared/src/test/require-jdk8/org/scalajs/testsuite/javalib/time/YearTest.scala [hunk 34]
```diff
@@ -0,0 +1,329 @@
...
+    assertTrue(min.isBefore(max))
+    assertFalse(max.isBefore(min))
+    assertFalse(max.isBefore(max))
+  }
+
+  @Test def now(): Unit = {
+    val now = LocalDate.now()
+    val year = Year.now()
+
+    assertEquals(now.getYear, year.getValue)
```
Comment (nicolasstucki) [inline test-suite/shared/src/test/require-jdk8/org/scalajs/testsuite/javalib/time/YearTest.scala]: This could fail if executed exactly on at new year. I would suggest to simply add a small message to the `assertEquals`: `assertEquals("Test for now failed on the current year (this test could fail if executed at new year).", now.getYear, year.getValue)`  or:

```scala
val now = LocalDate.now()
val year = Year.now()
if (now.getYear != year.getValue) {
  assertEquals(LocalDate.now(), year)
  println("Happy new year " + year + '!')
}

```

File: test-suite/shared/src/test/require-jdk8/org/scalajs/testsuite/javalib/time/YearTest.scala [hunk 35]
```diff
@@ -0,0 +1,329 @@
...
+  }
+
+  @Test def of(): Unit = {
+    assertEquals(lastBCYear, Year.of(0))
```
Comment (nicolasstucki) [inline test-suite/shared/src/test/require-jdk8/org/scalajs/testsuite/javalib/time/YearTest.scala]: Add a test for `Year.of(1)`

File: test-suite/shared/src/test/require-jdk8/org/scalajs/testsuite/javalib/time/YearTest.scala [hunk 36]
```diff
@@ -0,0 +1,329 @@
...
+    assertEquals(min, Year.of(MIN_VALUE))
+    assertEquals(max, Year.of(MAX_VALUE))
+  }
+
+  @Test def from(): Unit = {
+    for (t <- samples) {
+      assertSame(t, Year.from(t))
+    }
```
Comment (nicolasstucki) [inline test-suite/shared/src/test/require-jdk8/org/scalajs/testsuite/javalib/time/YearTest.scala]: Remove the braces.

File: test-suite/shared/src/test/require-jdk8/org/scalajs/testsuite/javalib/time/YearTest.scala [hunk 37]
```diff
@@ -0,0 +1,329 @@
...
+
+    val aDate = LocalDate.of(2015, 1, 1)
+    assertEquals(Year.of(2015), Year.from(aDate))
```
Comment (nicolasstucki) [inline test-suite/shared/src/test/require-jdk8/org/scalajs/testsuite/javalib/time/YearTest.scala]: Add tests for `min`, `max` and `lastBCYear`.

File: test-suite/shared/src/test/require-jdk8/org/scalajs/testsuite/javalib/time/YearTest.scala [hunk 38]
```diff
@@ -0,0 +1,329 @@
...
+  }
+
+  @Test def toStringOutput(): Unit = {
+    for (t <- samples) {
+      val expected = t.getValue().toString
+      assertEquals(expected, t.toString)
+    }
```
Comment (nicolasstucki) [inline test-suite/shared/src/test/require-jdk8/org/scalajs/testsuite/javalib/time/YearTest.scala]: Add a couple of tests with explicit strings:
```scala 
assertEquals("0", Year.of(0))
assertEquals("1", Year.of(1))
assertEquals("2016", Year.of(2016))
assertEquals("999999999", min))
```

**Comment by nicolasstucki** [comment]
That's all.

File: test-suite/shared/src/test/require-jdk8/org/scalajs/testsuite/javalib/time/YearTest.scala (hunk 30 above)
Comment (alonsodomin) [inline test-suite/shared/src/test/require-jdk8/org/scalajs/testsuite/javalib/time/YearTest.scala]: Shouldn't be 4 spaces identantion when breaking up a line in which several parameters are passed to a function (continuation line)?

File: test-suite/shared/src/test/require-jdk8/org/scalajs/testsuite/javalib/time/YearTest.scala (hunk 22 above)
Comment (alonsodomin) [inline test-suite/shared/src/test/require-jdk8/org/scalajs/testsuite/javalib/time/YearTest.scala]: Mind being a bit more specific what you mean by "unrolling" those loops?

File: test-suite/shared/src/test/require-jdk8/org/scalajs/testsuite/javalib/time/YearTest.scala (hunk 26 above)
Comment (alonsodomin) [inline test-suite/shared/src/test/require-jdk8/org/scalajs/testsuite/javalib/time/YearTest.scala]: Minus method is tested inherited from `TemporalTest`, it tests the method with a set of sample longs and verifies that it's outcome it's the same as using the `plus` method with negative input:

```scala
@Test def minus(): Unit = {
    for {
      temporal <- samples
      unit <- ChronoUnit.values if temporal.isSupported(unit)
      n <- sampleLongs
    } {
      testDateTime(temporal.minus(n, unit)) {
        if (n != Long.MinValue) temporal.plus(-n, unit)
        else temporal.plus(Long.MaxValue, unit).plus(1, unit)
      }
    }
  }
```

Didn't override it as none of the other `TemporalTest` implementors do so, they just test the concrete `minusXXX` methods (e.g.: `minusYears`)

**Comment by nicolasstucki** [comment]
Jenkins, retest this please.

**Comment by alonsodomin** [comment]
Added a fix to the `compareTo` test as it failed with the JDK implementation of `java.time.Year`

**Comment by nicolasstucki** [comment]
Jenkins, retest this please

File: javalib/src/main/scala/java/time/Year.scala [hunk 39]
```diff
@@ -0,0 +1,217 @@
...
+  // TODO
+  // def atMonth(month: Int): YearMonth = YearMonth.of(year, month)
+
+  def atMonthDay(monthDay: MonthDay): LocalDate = monthDay.atYear(year)
+
+  def compareTo(other: Year): Int = {
+    val diff = year - other.getValue
+    if (diff < 0) -1
+    else if (diff > 0) 1
+    else 0
```
Comment (nicolasstucki) [inline javalib/src/main/scala/java/time/Year.scala]: The result should be `diff` (`year - other.getValue`), in the REPL I'm getting:
```scala 
scala> Year.of(233).compareTo(Year.of(200))
res2: Int = 33

scala> Year.of(233).compareTo(Year.of(200))
res3: Int = 33

scala> Year.of(233).compareTo(Year.of(-200))
res4: Int = 433

scala> Year.of(233).compareTo(Year.of(999999999))
res5: Int = -999999766

scala> Year.of(-999999999).compareTo(Year.of(999999999))
res6: Int = -1999999998
```

File: test-suite/shared/src/test/require-jdk8/org/scalajs/testsuite/javalib/time/YearTest.scala [hunk 40]
```diff
@@ -0,0 +1,415 @@
...
+      assertEquals(LocalDate.now().getYear, year.getValue)
+      println("Happy new year " + year + '!')
+    }
+  }
+
+  @Test def of(): Unit = {
+    assertEquals(lastBCYear, Year.of(0))
+    assertEquals(firstACYear, Year.of(1))
+    assertEquals(min, Year.of(MIN_VALUE))
+    assertEquals(max, Year.of(MAX_VALUE))
```
Comment (nicolasstucki) [inline test-suite/shared/src/test/require-jdk8/org/scalajs/testsuite/javalib/time/YearTest.scala]: Should also test the contents of the year that is returned by `of`. Because if for some reason `Year.of(0)` would return `Year(5)` these tests would still pass.

**Comment by nicolasstucki** [comment]
That is all.

**Comment by alonsodomin** [comment]
Sorted

**Comment by nicolasstucki** [comment]
OK, it looks fine. It is time to squash the commits.

**Comment by nicolasstucki** [comment]
And change the commit message to `Add implementation for java.time.Year`.

**Comment by sjrd** [comment]
Jenkins, retest this please

**Comment by nicolasstucki** [comment]
LGTM
